### PR TITLE
Extension api

### PR DESCRIPTION
**Pull request by MMaas3**

### CODE REVIEW

File: src/main/java/nl/knaw/huc/resussun/dataextension/DataExtensionResponse.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                               Map<String, Map<String, List<? extends PropertyValue>>> rows) {
+    this.meta = meta;
+    this.rows = rows;
+  }
+
+  public interface PropertyValue {
+
+  }
+
+  public static class ResponseExtendProperty {
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/dataextension/DataExtensionResponse.java]: Additionally can have the property `type`:
``` 
{
    "id": "professionOrOccupation",
    "name": "Beruf oder Beschäftigung",
    "type": {
        "id": "SubjectHeading",
        "name": "Schlagwort"
    }
}
``` 
Are we going to support this?

File: src/main/java/nl/knaw/huc/resussun/dataextension/TimbuctooExtensionQuery.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      this.dataSetId = dataSetId;
+      this.subjects = subjects;
+      this.predicates = predicates;
+      this.escapeGraphQl = escapeGraphQl;
+    }
+
+    @Override
+    public Map<String, Map<String, List<? extends PropertyValue>>> mapResponse(JsonNode timbuctooResponse) {
+      JsonNode data = timbuctooResponse.get("data").get("dataSets").get(dataSetId);
+      Map<String, Map<String, List<? extends PropertyValue>>> mappedResponse = Maps.newHashMap();
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/dataextension/TimbuctooExtensionQuery.java]: Same comment as before: These Guava functions (newHasMap, newArrayList...) are considered deprecated. Just use `new HashMap<>();` and `new ArrayList<>();` instead.

File: src/main/java/nl/knaw/huc/resussun/dataextension/DataExtensionClient.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+          api.getDataSourceId(),
+          extensionRequest.getEntityIds(),
+          propertyNames
+      );
+      TimbuctooRequest timbuctooRequest = queryCreator.createQuery();
+      rows = timbuctoo.executeRequest(timbuctooRequest, queryCreator.createMapper());
+    }
+
+    List<ResponseExtendProperty> properties = extensionRequest.getProperties().stream().map(prop -> {
+      return new ResponseExtendProperty(prop.getId(), prop.getId(), prop.getSettings().orElse(null));
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/dataextension/DataExtensionClient.java]: Wrong name, should be something like:
`new ResponseExtendProperty(prop.getId(), prop.getName(), prop.getSettings().orElse(null));`

File: src/main/java/nl/knaw/huc/resussun/resources/ApiResource.java [hunk 4]
```diff
@@ -88,6 +127,26 @@ private ServiceManifest createServiceManifest(ApiData api) {
...
+                "Maximum number of values to return per row (0 for no limit)"
+            )).propertySetting(new ExtendPropertySetting<String>(
+                    "content",
+                    "Content",
+                    "select",
+                    "literal",
+                    "Content type: ID or literal"
+                ).choice("id", "ID")
+                 .choice("literal", "Literal")
+            )
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/resources/ApiResource.java]: These property settings are are declared in the model, but ignored if included in the actual request.

File: src/main/java/nl/knaw/huc/resussun/resources/DataExtensionPropertyProposalResource.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+      return Response.ok(new PropertyProposal(typeId, props)).build();
+
+    } catch (TimbuctooException e) {
+      LOG.error("Retrieve data from Timbuctoo failed", e);
+      return Response.serverError().entity("Could not retrieve metadata").build();
+    }
+  }
+
+  private static class PropertyProposal {
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/resources/DataExtensionPropertyProposalResource.java]: Should be in the `model` package?

File: src/main/java/nl/knaw/huc/resussun/timbuctoo/CollectionMetadata.java [hunk 6]
```diff
@@ -1,26 +1,57 @@
...
 
-import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.List;
+import java.util.Objects;
 
 public class CollectionMetadata {
-  @JsonProperty
-  private String name;
+  private final String collectionId;
+  private final String collectionListId;
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/timbuctoo/CollectionMetadata.java]: `collectionListId` is obsolete and as it cannot be mapped, every request fails.

File: src/main/java/nl/knaw/huc/resussun/timbuctoo/CollectionsMetadataMapper.java [hunk 7]
```diff
@@ -11,33 +11,76 @@
...
+  private static final String URI_PROP = "uri";
+  private final String key;
+
+  private CollectionsMetadataMapper(String key) {
+    this.key = key;
+  }
+
+  private static CollectionMetadata mapCollectionMetadata(JsonNode collection) {
+    String collectionId = collection.get(COLLECTION_ID_PROP).asText();
+    String collectionListId = collection.get(COLLECTION_LIST_ID_PROP).asText();
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/timbuctoo/CollectionsMetadataMapper.java]: There is no such property as `collectionListId` and therefore it fails. As it is not necessary anyway, we can just get rid of this line.

File: src/main/java/nl/knaw/huc/resussun/timbuctoo/Timbuctoo.java [hunk 8]
```diff
@@ -14,6 +14,7 @@
 public class Timbuctoo {
   private static final HttpClient HTTP_CLIENT = HttpClient.newBuilder().build();
   private final String url;
+  public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/timbuctoo/Timbuctoo.java]: Can be private

File: src/main/java/nl/knaw/huc/resussun/timbuctoo/TimbuctooRequest.java [hunk 9]
```diff
@@ -2,9 +2,8 @@
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 
-import java.util.Collections;
-import java.util.List;
 import java.util.Map;
+import java.util.Objects;
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/timbuctoo/TimbuctooRequest.java]: Import not used

File: src/main/java/nl/knaw/huc/resussun/ResussunApplication.java [hunk 10]
```diff
@@ -65,7 +66,7 @@ public void run(final ResussunConfiguration config, final Environment environmen
     final ApiClient apiClient = new ApiClient(redisConnection);
     environment.jersey().register(new ApiParamConverterProvider(apiClient));
     final UrlHelperFactory urlHelperFactory = config.getUrlHelperFactory();
-    environment.jersey().register(new ApiResource(searchClient, urlHelperFactory));
+    environment.jersey().register(new ApiResource(searchClient, urlHelperFactory, Timbuctoo::new));
```
[the remark this answers is not in the document]
Comment (MMaas3) [reply src/main/java/nl/knaw/huc/resussun/ResussunApplication.java]: This is the only solution how I could mock Timbuctoo in my tests.

File: src/main/java/nl/knaw/huc/resussun/dataextension/DataExtensionClient.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  public DataExtensionClient() {
+
+  }
+
+  public JsonNode createExtensionResponse(ApiData api, DataExtensionRequest extensionRequest)
+      throws TimbuctooException {
+    final Timbuctoo timbuctoo = api.getTimbuctoo();
+    List<String> propertyIds = extensionRequest.getProperties().stream().map(Property::getId).collect(
+        Collectors.toList());
+    Map<String, String> propretyIdNameMap = createPropertyIdNameMap(api, timbuctoo);
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/dataextension/DataExtensionClient.java]: Typo

File: src/main/java/nl/knaw/huc/resussun/model/DataExtensionResponse.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
+package nl.knaw.huc.resussun.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import nl.knaw.huc.resussun.model.ExtendPropertySetting;
```
Comment (kerim1) [inline src/main/java/nl/knaw/huc/resussun/model/DataExtensionResponse.java]: Unused import

File: src/test/java/nl/knaw/huc/resussun/resources/ApiParamConverterProviderMock.java [hunk 13]
```diff
@@ -1,28 +1,38 @@
 package nl.knaw.huc.resussun.resources;
 
+import nl.knaw.huc.resussun.ApiDataMock;
```
Comment (kerim1) [inline src/test/java/nl/knaw/huc/resussun/resources/ApiParamConverterProviderMock.java]: Unused import

File: src/test/java/nl/knaw/huc/resussun/resources/ApiParamConverterProviderMock.java [hunk 14]
```diff
@@ -1,28 +1,38 @@
...
 import nl.knaw.huc.resussun.api.ApiData;
+import nl.knaw.huc.resussun.timbuctoo.Timbuctoo;
```
Comment (kerim1) [inline src/test/java/nl/knaw/huc/resussun/resources/ApiParamConverterProviderMock.java]: Unused import

**Pull request merged**